File: ta4j-core/src/main/java/org/ta4j/core/rules/TrailingStopLossRule.java
/*
 * SPDX-License-Identifier: MIT
 */
package org.ta4j.core.rules;

import java.util.Objects;

import org.ta4j.core.BarSeries;
import org.ta4j.core.Indicator;
import org.ta4j.core.Position;
import org.ta4j.core.TradingRecord;
import org.ta4j.core.num.Num;

/**
 * A trailing stop-loss rule.
 *
 * <p>
 * Satisfied when the price reaches the trailing loss threshold.
 *
 * <p>
 * This rule uses the {@code tradingRecord}.
 */
public class TrailingStopLossRule extends AbstractRule implements StopLossPriceModel {

    /** The price indicator. */
    private final Indicator<Num> priceIndicator;

    /** The barCount. */
    private final int barCount;

    /** the loss-distance as percentage. */
    private final Num lossPercentage;

    /**
     * Constructor.
     *
     * @param indicator      the (close price) indicator
     * @param lossPercentage the loss percentage
     * @param barCount       the number of bars to look back for the calculation
     */
    public TrailingStopLossRule(Indicator<Num> indicator, Num lossPercentage, int barCount) {
        this(validatedConfig(indicator, lossPercentage, barCount));
    }

    private TrailingStopLossRule(Config config) {
        this.priceIndicator = config.priceIndicator();
        this.barCount = config.barCount();
        this.lossPercentage = config.lossPercentage();
    }

    private static Config validatedConfig(Indicator<Num> indicator, Num lossPercentage, int barCount) {
        Indicator<Num> validatedIndicator = Objects.requireNonNull(indicator, "priceIndicator");
        if (Num.isNaNOrNull(lossPercentage) || lossPercentage.isNegative()) {
            throw new IllegalArgumentException("lossPercentage must be >= 0");
        }
        if (barCount <= 0) {
            throw new IllegalArgumentException("barCount must be positive");
        }
        return new Config(validatedIndicator, lossPercentage, barCount);
    }

    /**
     * Constructor.
     *
     * @param indicator      the (close price) indicator
     * @param lossPercentage the loss percentage
     */
    public TrailingStopLossRule(Indicator<Num> indicator, Num lossPercentage) {
        this(validatedConfig(indicator, lossPercentage, Integer.MAX_VALUE));
    }

    /** This rule uses the {@code tradingRecord}. */
    @Override
    public boolean isSatisfied(int index, TradingRecord tradingRecord) {
        if (tradingRecord == null) {
            StopRuleTrace.traceUnavailable(this, index, "noTradingRecord");
            return false;
        }

        Position currentPosition = tradingRecord.getCurrentPosition();
        if (currentPosition == null || !currentPosition.isOpened() || currentPosition.getEntry() == null) {
            StopRuleTrace.traceUnavailable(this, index, "noOpenPosition");
            return false;
        }

        Num entryPrice = currentPosition.getEntry().getNetPrice();
        Num currentPrice = priceIndicator.getValue(index);
        int positionIndex = currentPosition.getEntry().getIndex();
        if (index < positionIndex) {
            StopRuleTrace.traceUnavailable(this, index, "indexBeforeEntry");
            return false;
        }
        boolean buy = currentPosition.getEntry().isBuy();
        int windowStartIndex = windowStartIndex(index, positionIndex);
        int lookback = index - windowStartIndex + 1;
        String extremeField = buy ? "highestPrice" : "lowestPrice";
        if (Num.isNaNOrNull(currentPrice)) {
            StopRuleTrace.traceTrailingDecision(this, index, false, buy, currentPrice, entryPrice, null, extremeField,
                    null, lookback, "lossPercentage", lossPercentage, "priceUnavailable");
            return false;
        }

        Num extremePrice;
        Num stopPrice;
        boolean satisfied;
        if (buy) {
            extremePrice = highestValue(windowStartIndex, index);
            if (Num.isNaNOrNull(extremePrice)) {
                StopRuleTrace.traceTrailingDecision(this, index, false, true, currentPrice, entryPrice, null,
                        extremeField, extremePrice, lookback, "lossPercentage", lossPercentage,
                        "extremePriceUnavailable");
                return false;
            }
            stopPrice = StopLossRule.stopLossPrice(extremePrice, lossPercentage, true);
            satisfied = currentPrice.isLessThanOrEqual(stopPrice);
        } else {
            extremePrice = lowestValue(windowStartIndex, index);
            if (Num.isNaNOrNull(extremePrice)) {
                StopRuleTrace.traceTrailingDecision(this, index, false, false, currentPrice, entryPrice, null,
                        extremeField, extremePrice, lookback, "lossPercentage", lossPercentage,
                        "extremePriceUnavailable");
                return false;
            }
            stopPrice = StopLossRule.stopLossPrice(extremePrice, lossPercentage, false);
            satisfied = currentPrice.isGreaterThanOrEqual(stopPrice);
        }
        String reason = satisfied ? "stopReached" : buy ? "priceAboveStop" : "priceBelowStop";
        StopRuleTrace.traceTrailingDecision(this, index, satisfied, buy, currentPrice, entryPrice, stopPrice,
                extremeField, extremePrice, lookback, "lossPercentage", lossPercentage, reason);
        return satisfied;
    }

    /**
     * Returns the stop-loss price for the supplied position entry.
     *
     * @param series   the price series
     * @param position the position being evaluated
     * @return the stop-loss price, or {@code null} if unavailable
     * @since 0.22.3
     */
    @Override
    public Num stopPrice(BarSeries series, Position position) {
        if (position == null || position.getEntry() == null) {
            return null;
        }
        // If the entry bar was evicted by max-bar-count retention, fall back to
        // the first retained bar as the anchor (an approximation of entry-time stop).
        int entryIndex = retainedStartIndex(position.getEntry().getIndex());
        if (position.getEntry().isBuy()) {
            Num highestCloseNum = priceIndicator.getValue(entryIndex);
            if (Num.isNaNOrNull(highestCloseNum)) {
                return null;
            }
            return StopLossRule.stopLossPrice(highestCloseNum, lossPercentage, true);
        }
        Num lowestCloseNum = priceIndicator.getValue(entryIndex);
        if (Num.isNaNOrNull(lowestCloseNum)) {
            return null;
        }
        return StopLossRule.stopLossPrice(lowestCloseNum, lossPercentage, false);
    }

    private int windowStartIndex(int index, int positionIndex) {
        int activeBarCount = Math.min(index - positionIndex + 1, barCount);
        return retainedStartIndex(index - activeBarCount + 1);
    }

    private int retainedStartIndex(int requestedStartIndex) {
        return Math.max(requestedStartIndex, priceIndicator.getBarSeries().getBeginIndex());
    }

    private Num highestValue(int startIndex, int endIndex) {
        Num highest = null;
        for (int index = startIndex; index <= endIndex; index++) {
            Num candidate = priceIndicator.getValue(index);
            if (!Num.isNaNOrNull(candidate) && (highest == null || candidate.isGreaterThan(highest))) {
                highest = candidate;
            }
        }
        return highest;
    }

    private Num lowestValue(int startIndex, int endIndex) {
        Num lowest = null;
        for (int index = startIndex; index <= endIndex; index++) {
            Num candidate = priceIndicator.getValue(index);
            if (!Num.isNaNOrNull(candidate) && (lowest == null || candidate.isLessThan(lowest))) {
                lowest = candidate;
            }
        }
        return lowest;
    }

    private record Config(Indicator<Num> priceIndicator, Num lossPercentage, int barCount) {
    }
}

File: ta4j-core/src/main/kotlin/org/ta4j/core/rules/TrailingStopLossRule.kt
/**
 * The MIT License (MIT)
 *
 * Copyright (c) 2017-2022 Ta4j Organization & respective
 * authors (see AUTHORS)
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy of
 * this software and associated documentation files (the "Software"), to deal in
 * the Software without restriction, including without limitation the rights to
 * use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies of
 * the Software, and to permit persons to whom the Software is furnished to do so,
 * subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS
 * FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR
 * COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER
 * IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN
 * CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
 */
package org.ta4j.core.rules

import org.ta4j.core.Indicator
import org.ta4j.core.TradingRecord
import org.ta4j.core.indicators.helpers.HighestValueIndicator
import org.ta4j.core.indicators.helpers.LowestValueIndicator
import org.ta4j.core.num.*
import kotlin.math.min

/**
 * A trailing stop-loss rule
 *
 * Satisfied when the price reaches the trailing loss threshold.
 */
class TrailingStopLossRule
/**
 * Constructor.
 *
 * @param priceIndicator      the (close price) indicator
 * @param lossPercentage the loss percentage
 */ @JvmOverloads constructor(
    /**
     * The price indicator
     */
    private val priceIndicator: Indicator<Num>,
    /** the loss-distance as percentage  */
    private val lossPercentage: Num,
    /** The barCount  */
    private val barCount: Int = Int.MAX_VALUE
) : AbstractRule() {
    /**
     * the current stop loss price activation
     */
    var currentStopLossLimitActivation: Num? = null
        private set
    /**
     * Constructor.
     *
     * @param priceIndicator      the (close price) indicator
     * @param lossPercentage the loss percentage
     * @param barCount       number of bars to look back for the calculation
     */
    /** This rule uses the `tradingRecord`.  */
    override fun isSatisfied(index: Int, tradingRecord: TradingRecord?): Boolean {
        var satisfied = false
        // No trading history or no position opened, no loss
        if (tradingRecord != null) {
            val currentPosition = tradingRecord.getCurrentPosition()
            if (currentPosition.isOpened) {
                val currentPrice = priceIndicator[index]
                val positionIndex = currentPosition.entry!!.index
                satisfied = if (currentPosition.entry!!.isBuy) {
                    isBuySatisfied(currentPrice, index, positionIndex)
                } else {
                    isSellSatisfied(currentPrice, index, positionIndex)
                }
            }
        }
        traceIsSatisfied(index, satisfied)
        return satisfied
    }

    private fun isBuySatisfied(currentPrice: Num?, index: Int, positionIndex: Int): Boolean {
        val highest = HighestValueIndicator(
            priceIndicator,
            getValueIndicatorBarCount(index, positionIndex)
        )
        val highestCloseNum = highest[index]
        val lossRatioThreshold =
            highestCloseNum.numOf(100).minus(lossPercentage).div(highestCloseNum.numOf(100))
        currentStopLossLimitActivation = highestCloseNum.times(lossRatioThreshold)
        return currentPrice!!.isLessThanOrEqual(currentStopLossLimitActivation)
    }

    private fun isSellSatisfied(currentPrice: Num?, index: Int, positionIndex: Int): Boolean {
        val lowest = LowestValueIndicator(
            priceIndicator,
            getValueIndicatorBarCount(index, positionIndex)
        )
        val lowestCloseNum = lowest[index]
        val lossRatioThreshold = lowestCloseNum.numOf(100).plus(lossPercentage).div(lowestCloseNum.numOf(100))
        currentStopLossLimitActivation = lowestCloseNum.times(lossRatioThreshold)
        return currentPrice!!.isGreaterThanOrEqual(currentStopLossLimitActivation)
    }

    private fun getValueIndicatorBarCount(index: Int, positionIndex: Int): Int {
        return min(index - positionIndex + 1, barCount)
    }

    override fun traceIsSatisfied(index: Int, isSatisfied: Boolean) {
        if (log.isTraceEnabled) {
            log.trace(
                "{}#isSatisfied({}): {}. Current price: {}, Current stop loss activation: {}",
                javaClass.simpleName, index, isSatisfied, priceIndicator[index],
                currentStopLossLimitActivation
            )
        }
    }
}